### ISSUE TITLE
Preprocess data for inference on Graph2Tree

### ISSUE THREAD

**Issue by maurovit** [graph4ai/graph4nlp#319]
Hello everyone,
I want to ask something that is related to https://github.com/graph4ai/graph4nlp/issues/316#issue-918894490.
I cannot fully understand how I can pass the data to the model for inference. 
In particular, for the evaluation in tutorial Graph2Tree about jobs data, I see that the model retrieves the data from the test data-loader, which is generate from JobsDatasetForTree, which gets data from text file and preprocess them. 
This is different from my goals: infact, after training, I save the model and I load it, after that I want to do inference on unseen data; so, I want to pass to model a list of strings, preprocess these strings (according to the method that I used in the training for the graph construction, i.e. Dependency or Constituency or something else) and do inference on this data. 
Can you explain me how i can do that? Thanks

**Comment by AlanSwift** [comment]
During the inference stage, for unseen data, you can build another Dataloader (just as the training Dataloader) and pass it to the model.

**Comment by maurovit** [comment]
Yes, that's true but the point is that i should get from dataloader these following items
`eval_input_graph, batch_tree_list, batch_original_tree_list = data['graph_data'], data['dec_tree_batch'], data['original_dec_tree_batch']`
How can I create that data?

**Comment by AlanSwift** [comment]
Actually, it is automatically created when you build the dataloader. Please refer to the [code](https://github.com/graph4ai/graph4nlp/blob/master/examples/pytorch/semantic_parsing/graph2tree/jobs/src/runner.py#L129) in the example.
Specifically, when you initialize the instance ``self.test_data_loader``, it will read the raw text stored in [``raw_file_names``](https://github.com/graph4ai/graph4nlp/blob/a53bd803b5f85cd479b371e31940bc029e33ccb9/graph4nlp/pytorch/datasets/jobs.py#L90) and automatically build the data you need.

**Comment by AlanSwift** [comment]
More details:
For your own test data, you can write them in a single file: "test.txt" with the format as follows: Each line is a sample and each sample is "A\tB" (A is input and B is output). Thus it can be parsed by the default parsing function in [``Text2TextDataset.parse_file``](https://github.com/graph4ai/graph4nlp/blob/a53bd803b5f85cd479b371e31940bc029e33ccb9/graph4nlp/pytorch/data/dataset.py#L696). (If you want to use your own format, you should override this parsing function in your own dataset class).
We will make a more clear explanation in the document for this part in the next version.

**Comment by maurovit** [comment]
First of all, thank you for your reply. 
But if I want to replicate this process for a single sentence, or a list of sentences, without using any text file? Is not possibile?
What I am asking for is a method for preprocess the input in the inference phase, not during evaluation phase. So for example, i want to input a sentence to the model and get the prediction. The problem is that i am stuck on the preprocessing of this sentence to pass the sentence preprocessed to the model. During inference phase I have only the input, not the output, so the format "A\tB" is not respected.

**Comment by maurovit** [comment]
Maybe my explanation is not clear. I had trained the model, and i had evaluated it against test data after that I saved it. Now I loaded this model and I want to use it to make prediction about unseen data (so data with no label) and I want to preprocess the sentences in input in the same way that i did during the evaluation/training stage. The problem is that the process is automated for the data that are in the train.txt or test.txt files but there is no method do repeat this process for a string or a list of strings during prediction stage

**Comment by AlanSwift** [comment]
Currently, we are sorry that we don't support this feature.

But you can implement it by modifying the Text2TextDataloader's "read in" without too much effort.

A simple idea is to override the function ["read_raw_data"](https://github.com/graph4ai/graph4nlp/blob/a53bd803b5f85cd479b371e31940bc029e33ccb9/graph4nlp/pytorch/data/dataset.py#L400) according to your need. Maybe you can read in a simple string-like input and store it in ``self.test``. As for the output, since it is not necessary during the test stage, you can just use a dummy string (e.g., "asdf...") for simplicity or totally remove it during the inference pipeline.

**Comment by maurovit** [comment]
Got it. According to you, may I use the static function `_build_topology_process` to create the input graph for the network?

**Comment by AlanSwift** [comment]
Yes, it is part of building the graph.
Note that there are other necessary steps (e.g., vectorization). If you want to decouple the training and inference stages, maybe you should read the code in [``_preprocess()``](https://github.com/graph4ai/graph4nlp/blob/a53bd803b5f85cd479b371e31940bc029e33ccb9/graph4nlp/pytorch/data/dataset.py#L659).
For simplicity, if you are not sure about how to decouple, I suggest you: 1. use our pipeline and write your own test cases in ``test.txt`` to replace our default test file. 2. keep the "train.txt" unchanged. 3. delete the original processed folder. Then it would work as you want without too much effort.

For example, you have input "abcdefg" and no output, you can save it to a file (test.txt) with "abcdefg\tNOOUTPUT".

**Comment by schenglee** [comment]
Thank you @maurovit ,  and @AlanSwift  is right. 
We will refine the inference interface of the model in the next release.  
For a temporary solution, I recommend you to append some inference string lines like ```the string input \t <TBD>``` to the ```test.txt``` file, and skip them in the evaluation process. You can translate them after you have complete training and model selection.  A example can be found in this branch ```add_inference_example_graph2tree```.

**Comment by maurovit** [comment]
Thank you all, your suggestions are very precious. 🙏

**Comment by AlanSwift** [comment]
Ok, I will close this issue, and feel free to ask us any questions.
Thank you again for your feedback. We will discuss this problem in our group meeting and improve it in the next version.
